File: apps/server/src/game/invites.ts
import { Color, type Mode, type TimeControl } from "@chesstalk/shared/enums";
import type { OpponentInfo } from "@chesstalk/shared/wire";
import { randomUUID } from "node:crypto";
import type { WebSocket } from "ws";
import { GameActor } from "./game-actor.ts";

interface Invite {
  userId: string;
  ws: WebSocket;
  opponentInfo: OpponentInfo;
  mode: Mode;
  timeControl: TimeControl;
}

export interface PendingInvite {
  userId: string;
  mode: Mode;
  timeControl: TimeControl;
}

interface PairedSide {
  userId: string;
  ws: WebSocket;
  color: Color;
  opponent: OpponentInfo;
}

export interface InviteMatchResult {
  matched: true;
  game: GameActor;
  self: PairedSide;
  other: PairedSide;
}

export type JoinInviteResult =
  | { matched: false; reason: "not_found" | "self_join" }
  | InviteMatchResult;

export class InviteRegistry {
  private readonly invites: Map<string, Invite> = new Map();
  private readonly userToInvite: Map<string, string> = new Map();

  create(
    userId: string,
    ws: WebSocket,
    mode: Mode,
    timeControl: TimeControl,
    opponentInfo: OpponentInfo,
  ): string {
    this.leaveByUser(userId);

    const inviteId = randomUUID();
    this.invites.set(inviteId, {
      userId,
      ws,
      opponentInfo,
      mode,
      timeControl,
    });
    this.userToInvite.set(userId, inviteId);
    return inviteId;
  }

  peek(inviteId: string): PendingInvite | null {
    const invite = this.invites.get(inviteId);
    if (!invite) return null;
    return {
      userId: invite.userId,
      mode: invite.mode,
      timeControl: invite.timeControl,
    };
  }

  join(
    inviteId: string,
    userId: string,
    ws: WebSocket,
    opponentInfo: OpponentInfo,
  ): JoinInviteResult {
    const invite = this.invites.get(inviteId);
    if (!invite) return { matched: false, reason: "not_found" };
    if (invite.userId === userId) return { matched: false, reason: "self_join" };

    this.invites.delete(inviteId);
    this.userToInvite.delete(invite.userId);
    this.leaveByUser(userId);

    const now = Date.now();
    const joinerIsWhite = Math.random() < 0.5;
    const whiteSide = joinerIsWhite
      ? { userId, ws, info: opponentInfo }
      : { userId: invite.userId, ws: invite.ws, info: invite.opponentInfo };
    const blackSide = joinerIsWhite
      ? { userId: invite.userId, ws: invite.ws, info: invite.opponentInfo }
      : { userId, ws, info: opponentInfo };

    const game = new GameActor({
      id: randomUUID(),
      mode: invite.mode,
      timeControl: invite.timeControl,
      white: {
        userId: whiteSide.info.userId,
        username: whiteSide.info.username,
        ratingBefore: whiteSide.info.rating,
        ratingAfter: null,
      },
      black: {
        userId: blackSide.info.userId,
        username: blackSide.info.username,
        ratingBefore: blackSide.info.rating,
        ratingAfter: null,
      },
      now,
    });

    const joinerColor: Color = joinerIsWhite ? Color.White : Color.Black;
    const inviterColor: Color = joinerIsWhite ? Color.Black : Color.White;

    return {
      matched: true,
      game,
      self: {
        userId,
        ws,
        color: joinerColor,
        opponent: invite.opponentInfo,
      },
      other: {
        userId: invite.userId,
        ws: invite.ws,
        color: inviterColor,
        opponent: opponentInfo,
      },
    };
  }

  leaveByUser(userId: string): void {
    const inviteId = this.userToInvite.get(userId);
    if (!inviteId) return;
    const invite = this.invites.get(inviteId);
    if (invite?.userId === userId) {
      this.invites.delete(inviteId);
    }
    this.userToInvite.delete(userId);
  }
}

export const inviteRegistry = new InviteRegistry();

File: apps/server-go/internal/game/invites.go
package game

import (
	"sync"
	"time"
)

type Invite struct {
	UserID       string
	Sender       Sender
	OpponentInfo OpponentInfo
	Mode         string
	TimeControl  TimeControl
	CreatedAt    time.Time
}

type PendingInvite struct {
	UserID      string
	Mode        string
	TimeControl TimeControl
}

type InviteJoinResult struct {
	Matched bool
	Reason  string
	Game    *Actor
	Self    PairedSide
	Other   PairedSide
}

type InviteRegistry struct {
	mu           sync.Mutex
	invites      map[string]Invite
	userToInvite map[string]string
}

func NewInviteRegistry() *InviteRegistry {
	return &InviteRegistry{
		invites:      make(map[string]Invite),
		userToInvite: make(map[string]string),
	}
}

func (r *InviteRegistry) Create(userID string, sender Sender, mode string, tc TimeControl, info OpponentInfo) string {
	r.mu.Lock()
	defer r.mu.Unlock()

	r.leaveByUserLocked(userID)
	inviteID := randomID()
	r.invites[inviteID] = Invite{
		UserID:       userID,
		Sender:       sender,
		OpponentInfo: info,
		Mode:         mode,
		TimeControl:  tc,
		CreatedAt:    time.Now(),
	}
	r.userToInvite[userID] = inviteID
	return inviteID
}

func (r *InviteRegistry) Peek(inviteID string) (PendingInvite, bool) {
	r.mu.Lock()
	defer r.mu.Unlock()
	invite, ok := r.invites[inviteID]
	if !ok {
		return PendingInvite{}, false
	}
	return PendingInvite{UserID: invite.UserID, Mode: invite.Mode, TimeControl: invite.TimeControl}, true
}

func (r *InviteRegistry) Join(inviteID string, userID string, sender Sender, info OpponentInfo) InviteJoinResult {
	r.mu.Lock()
	defer r.mu.Unlock()

	invite, ok := r.invites[inviteID]
	if !ok {
		return InviteJoinResult{Reason: "not_found"}
	}
	if invite.UserID == userID {
		return InviteJoinResult{Reason: "self_join"}
	}

	delete(r.invites, inviteID)
	delete(r.userToInvite, invite.UserID)
	r.leaveByUserLocked(userID)

	newcomer := &Waiter{
		UserID:       userID,
		Sender:       sender,
		OpponentInfo: info,
		Mode:         invite.Mode,
		TimeControl:  invite.TimeControl,
	}
	peer := &Waiter{
		UserID:       invite.UserID,
		Sender:       invite.Sender,
		OpponentInfo: invite.OpponentInfo,
		Mode:         invite.Mode,
		TimeControl:  invite.TimeControl,
	}
	result := pairPlayers(newcomer, peer, invite.Mode, invite.TimeControl)

	return InviteJoinResult{
		Matched: true,
		Game:    result.Game,
		Self:    result.Self,
		Other:   result.Other,
	}
}

func (r *InviteRegistry) LeaveByUser(userID string) {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.leaveByUserLocked(userID)
}

func (r *InviteRegistry) leaveByUserLocked(userID string) {
	inviteID := r.userToInvite[userID]
	if inviteID == "" {
		return
	}
	invite, ok := r.invites[inviteID]
	if ok && invite.UserID == userID {
		delete(r.invites, inviteID)
	}
	delete(r.userToInvite, userID)
}
